db/view/view.cc

```diff
@@ -3026,7 +3026,130 @@ void view_builder::init_virtual_table() {
+        inject_failure("view_builder_flush_fragments");
+        check_abort_source();
+        if (!_fragments.empty()) {
+            _fragments.emplace_front(*reader().schema(), permit(), partition_start(get_current_key(), tombstone()));
+            auto base_schema = base()->schema();
+            auto views = with_base_info_snapshot(_views_to_build);
+            auto fragemnts_reader = make_mutation_reader_from_fragments(reader().schema(), permit(), std::move(_fragments));
+            auto close_reader = defer([&fragemnts_reader] { fragemnts_reader.close().get(); });
+            reader().upgrade_schema(base_schema);
+            _gen->populate_views(
+                    *base(),
+                    std::move(views),
+                    get_current_key().token(),
+                    std::move(fragemnts_reader),
+                    _now).get();
+            close_reader.cancel();
+            _fragments.clear();
+            _fragments_memory_usage = 0;
+        }
+    }
+
+public:
+    view_consumer(shared_ptr<view_update_generator> gen, gc_clock::time_point now, std::vector<view_ptr> views_to_build)
+        : _gen(std::move(gen))
+        , _now(now)
+        , _views_to_build(std::move(views_to_build))
+        {}
+
+    stop_iteration consume_new_partition(const dht::decorated_key& dk) {
+        inject_failure("view_builder_consume_new_partition");
+        if (dk.key().is_empty()) {
+            on_internal_error(vlogger, format("Trying to consume empty partition key {}", dk));
+        }
+        set_current_key(std::move(dk));
+        bool stop = setup_new_partition();
+        return stop_iteration(stop);
+    }
+
+    stop_iteration consume(tombstone) {
+        inject_failure("view_builder_consume_tombstone");
+        return stop_iteration::no;
+    }
+
+    stop_iteration consume(static_row&& sr, tombstone, bool) {
+        inject_failure("view_builder_consume_static_row");
+        if (stop_consuming()) {
+            return stop_iteration::yes;
+        }
+
+        add_fragment(std::move(sr));
+        return stop_iteration::no;
+    }
+
+    stop_iteration consume(clustering_row&& cr, row_tombstone, bool is_live) {
+        inject_failure("view_builder_consume_clustering_row");
+        if (!is_live) {
+            return stop_iteration::no;
+        }
+        if (stop_consuming()) {
+            return stop_iteration::yes;
+        }
+
+        add_fragment(std::move(cr));
+        return stop_iteration::no;
+    }
+
+    stop_iteration consume(range_tombstone_change&&) {
+        inject_failure("view_builder_consume_range_tombstone");
+        return stop_iteration::no;
+    }
+
+    stop_iteration consume_end_of_partition() {
+        inject_failure("view_builder_consume_end_of_partition");
+        flush_fragments();
+        return stop_iteration(stop_consuming_end_of_partition());
+    }
+};
+
+// Called in the context of a seastar::thread.
+class view_builder::consumer : public view_consumer {
```

Comment:
Nit: if the inheritance is only used for the sake of code reuse, you can consider using private inheritance instead of public.

Alternatively, we could consider going with a CRTP approach to avoid dynamic dispatch - I'm not sure it is worth it, though (and if it is then it belongs to a follow-up).

Reply:
It cannot be private inheritance because `view_builder::consumer` needs to satisfy `CompactedFragmentsConsumerV2` which requires for instance `obj.consume_new_partition(dk)`.
With private inheritance, the method is private, so the concept is not satisfied.

Reply:
Ok, I see. In that case it looks like CRTP would work, but it's not worth doing the transformation right now.